infisical: support secrets within paths for `data` references

lgo opened the pull request:
(cc @akhilmhdh)

## Problem Statement

When attempting to access a secret in a directory, provided `secretsPath: /`, like so:
```yaml
apiVersion: external-secrets.io/v1beta1
kind: ExternalSecret
metadata:
  name: infisical-managed-secrets
spec:
  secretStoreRef:
    kind: SecretStore
    name: infisical
  target:
    name: auth-api
  data:
    - secretKey: API_KEY
      remoteRef:
        key: "foo/bar"
``` 

This will always result in a 404 API error from Infisical, because the API request is malformed (`/api/v3/secrets/raw/foo/bar`). The correct API request would instead be `/api/v3/secrets/raw/bar?secretPath=/foo`. This either:
- Made it so you simply cannot use secrets in folders with direct references via `data`
- Made it so you have to use a separate secret provider with a distinct `path`, which isn't very usable or very obvious

In combination with #4059, this would result in empty secrets but after that issue is fixed these would fail to sync with an error.

Overall this was also quite confusing with the `SecretStore`'s `secretsPath` property - would a `remoteRef` be an additive path? Would it supercede the `SecretStore`?


## Related Issue

Fixes #4298

## Proposed Changes

As suggested on the issue (https://github.com/external-secrets/external-secrets/issues/4298#issuecomment-2600966462) this makes a particular changes:
- `remoteRef`'s with paths take precedence over any `secretsPath` on the `SecretStore`
- If a `remoteRef` is detected to be a path (i.e. it contains `/`), an error is raised if it does not start with `/` with an actionable error indicating to use `/` and that it is an absolute reference. This should hopefuly further clarify the interaction between this, and `secretsPath`
- Documentation is updated with more specific examples of data in Infisical and how you would retrieve it, now that these features work

I've tested this against a self-hosted Infisical instance. I used a branch rebased off of #4304 to test this ([lgo:joey-infisical-folder-handling-testing](https://github.com/lgo/external-secrets/tree/joey-infisical-folder-handling-testing))
- With a `remoteRef: {key: "foo/bar"}`
- :x: `ExternalSecret` expectedly fails with `a secret key referencing a folder must start with a '/' as it is an absolute path, key: foo/bar`
- With a `remoteKey: {key: "/foo/bar"}`
- ✅ the secret successfully syncs.

**TODO**:
- [x] This needs to be rebased after merging #4304. I had a few changes layered on that I'll include after the rebase.

## Checklist

- [x] I have read the [contribution guidelines](https://external-secrets.io/latest/contributing/process/#submitting-a-pull-request)
- [x] All commits are signed with `git commit --signoff`
- [ ] My changes have reasonable test coverage
- [x] All tests pass with `make test`
- [x] I ensured my PR is ready for review with `make reviewable`

akhilmhdh commented:
Hi @lgo 

That was quick. Code wise looking good to me. I'll do the application test today and give my approval on my side. 

One small issue is, can we bring back the secret path option. Removing it would be a breaking change and want to keep it backward compatibility as much as we can. Simply keep that option as it is and if user provides and not `defaultSecretPath`, it will be used. The docs can be kept with new one, may be mentioning `secretPath is depreciated` would be great. As both option was optional from the start, this should give customers less headache.

lgo commented:
> One small issue is, can we bring back the secret path option. Removing it would be a breaking change and want to keep it backward compatible. Simply keep that option as it is and if user provides it and not defaultSecretPath, the first one will be used. The docs can be kept with new one, may be mentioning secretPath is depreciated would be great. As both option was optional from the start, this should give customers less headache.

Sounds good. We could also just keep `secretPath` and ensure the documentation is clear. Your suggestion of keeping it around if we do add `defaultSecretPath` is a much better idea than changing it.

lgo commented:
I decided to just avoid adding/changing `defaultSecretPath` and pushed up the changes, a few test fixes, and quality check fixes. This should now be good and just has a few test conflicts with the other PR (which I'll address once it's in).

Skarlso commented:
@lgo Hello 👋  Could you please sign your commits? :)

lgo commented:
Done! Sorry, I was away on a trip. I have not yet had a chance to build out the e2e tests either.

Hi @akhilmhdh - just checking if you had any additional testing you'd like to do on your end before checking off?

Skarlso commented:
Is this now ready for review?

Skarlso commented on pkg/provider/infisical/client.go:
```diff
@@ -43,14 +43,44 @@ func getPropertyValue(jsonData, propertyName, keyName string) ([]byte, error) {
+
+	// Otherwise, take the prefix from `key` and use that as the path. We intentionally discard
+	// `defaultPath`.
+	keyParts := strings.Split(key, "/")
+	keySuffix := keyParts[len(keyParts)-1]
+	path := strings.Join(keyParts[:len(keyParts)-1], "/")
+	return path, keySuffix, nil
```
```suggestion
	lastIndex := strings.LastIndex(key, "/")
	return key[lastIndex+1:], key[:lastIndex], nil
```

Skarlso commented on pkg/provider/infisical/provider.go:
```diff
@@ -95,13 +95,18 @@ func (p *Provider) NewClient(ctx context.Context, store esv1beta1.GenericStore,
 
+		secretPath := infisicalSpec.SecretsScope.SecretsPath
+		if secretPath == "" {
+			secretPath = "/"
```
Won't this break existing things? Like, if you expected this path not to be set previously, it will be defaulted now to `/`.

lgo replied:
I need to recall why exactly I added this, but I think it was necessary?

I believe ExternalSecrets didn't work at all if you pass a secret name that includes a `/`, so you can always assume the only secrets that worked were from the root (`/`) and I believe this cannot be simply `""` when making an API request but I'm doing a quick pass to figure out why.

lgo replied:
Hmm I guess that's not necessarily true as `getSecretAddress` will return `("", "FOOBAR")` when the path is empty or has a `/`. I'll have to deploy this and test it.

Skarlso replied:
I verified the error. 👍

gusfcarvalho commented:
Hi @lgo ! Were you able to address the points @Skarlso mentioned?

akhilmhdh commented:
Hi @Skarlso @gusfcarvalho @lgo 

I don't have push access to this branch. Can I make the changes as another PR with your commits to make this merged, if @lgo doesn't have time to continue working on it? We would love to see this changes getting shipped.

gusfcarvalho commented:
As long as you’re fork from @lgo repo, so his contributions are correctly accredited to him, I’m up for it 😄

lgo commented:
Heh, yea - sorry this went stale! I haven't had a lot of spare time lately and I've worked around this by using a custom build that included this patch.

Do feel free to fork this to take it over the finish line (if you don't get to it by EOW, I do have a long weekend that I can likely complete this!)

lgo commented:
(actually, I'll go ahead and handle the rebase right now and the existing PR feedback just to leave a slightly cleaner slate, whether or not there's further work needed)

lgo commented:
Okay I've pushed up a rebase, squashed the commits down, and addressed the one feedback. Two things that still need to be done:
- [ ] Deploy this and step through the test cases I outlined again
- [ ] Check whether Infisical's API requires a leading `/` on the `secretPath` query param (https://infisical.com/docs/api-reference/endpoints/secrets/read#parameter-secret-path). If they do, I'll make this much clearer in the code as I've partially written up error / defensive cases for tihs.

typedrat commented:
> Okay I've pushed up a rebase, squashed the commits down, and addressed the one feedback. Two things that still need to be done:
> 
>     * [ ]  Make sure the unit tests are still working. (For some reason `go test <file>` isn't working and `make unit-test` is failing but with a large wall of text so I need to revisit that)
> 
>     * [ ]  Deploy this and step through the test cases I outlined again
> 
>     * [ ]  Check whether Infisical's API requires a leading `/` on the `secretPath` query param (https://infisical.com/docs/api-reference/endpoints/secrets/read#parameter-secret-path). If they do, I'll make this much clearer in the code as I've partially written up error / defensive cases for tihs.
> 
> 
> I'll take a pass at this later tonight, or tomorrow morning, and see if I can wrap these items up.

Any updates on this? I really prefer using `external-secrets` over the native Infisical operator, and this would really help make it a lot more usable.

gusfcarvalho commented:
Hi @typedrat! If you can take the remaining points from these commits and push your contribution to it, I’d gladly review it!

otherwise, chances are this might go stale just because that’s the nature of community driven effort 🤷🏽‍♂️

typedrat commented:
> Hi @typedrat! If you can take the remaining points from these commits and push your contribution to it, I’d gladly review it!
> 
> otherwise, chances are this might go stale just because that’s the nature of community driven effort 🤷🏽‍♂️

Fair enough, I know how open source goes haha

I don't know Go and I've only got the barest bit of experience with using Infisical's API, so I don't know if I'm really the one to get this over the finish line myself, but I'll come back and look at this when I'm out of my work crunch if no one else has picked this up.

Skarlso commented:
@typedrat I'll take a look. :)

Skarlso commented:
I had to rebase to get the signed commit to work, but I left an `On-behalf-of` in the commit message to make sure to credit OP. :)

Adriien-M commented:
Thanks @Skarlso for your work 👍
